packit/upstream.py

```diff
@@ -305,41 +304,91 @@ def create_archive(
+        List tags in the repository sorted by created date from the most recent.
         """
-
-        logger.debug(
-            f"We're about to `git-describe` the upstream repository "
-            f"{self.local_project.working_dir}."
-        )
-
         try:
-            version_tag_glob = self.package_config.upstream_tag_template.replace(
-                "{version}", "?*"
-            )
             cmd = [
                 "git",
-                "describe",
-                "--tags",
-                "--abbrev=0",
-                f"--match={version_tag_glob}",
+                "tag",
+                "--list",
+                "--sort=-creatordate",
             ]
-            if before:
-                cmd += [f"{before}^"]
-            last_tag = run_command(
+            tags = run_command(
                 cmd,
                 output=True,
                 cwd=self.local_project.working_dir,
-            ).stdout.strip()
+            ).stdout.split()
         except PackitCommandFailedError as ex:
             logger.debug(f"{ex!r}")
-            logger.info("Can't describe this repository, are there any git tags?")
-            # no tags in the git repo
+            logger.info("Can't list the tags in this repository.")
+            return []
+
+        return tags
+
+    def get_last_tag(
+        self, matching_only: bool = False, before: str = None
+    ) -> Optional[str]:
+        """
+        Get last git-tag from the repo.
+
+        Args:
+            matching_only: whether to consider the `upstream_tag_include` and
+                `upstream_tag_exclude` regexes to filter out matching tags.
+            before: get the last tag before this tag
+        """
+
+        logger.debug(
+            f"We're about to get latest {'matching' if matching_only else ''} tag in "
+            f"the upstream repository {self.local_project.working_dir}."
+        )
+
+        tags = self.list_tags()
+
+        if not tags:
+            logger.info("No tags found in the repository.")
             return None
-        return last_tag
+
+        if before:
+            if before not in tags:
+                logger.debug(f"{before} not present in the obtained list of tags.")
+                return None
+            index = tags.index(before)
+            tags = tags[(index + 1) :]  # noqa
+
+        matching_tags = self.filter_tags(tags) if matching_only else tags
+        return matching_tags[0] if matching_tags else None
+
+    def filter_tags(self, tags: List[str]):
+        """
+        Filter the given tags using `upstream_tag_include` and
+        `upstream_tag_exclude` if they are present.
+
+        Args:
+            tags: list of tags that should be filtered
+        """
+        if self.package_config.upstream_tag_include:
+            tags = [
+                tag
+                for tag in tags
+                if re.match(self.package_config.upstream_tag_include, tag)
```

Comment:
Since this is called in a loop, it would be more efficient to `re.compile()` the regex first.

Reply:
good point!